<!DOCTYPE html>

<html lang="en" data-content_root="../../../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>maayanlab_bioinformatics.enrichment.gsea2005 &#8212; Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nature.css?v=601dbdee" />
    <script src="../../../_static/documentation_options.js?v=d45e8c67"></script>
    <script src="../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">maayanlab_bioinformatics.enrichment.gsea2005</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for maayanlab_bioinformatics.enrichment.gsea2005</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<div class="viewcode-block" id="GSEA2005">
<a class="viewcode-back" href="../../../maayanlab_bioinformatics.enrichment.html#maayanlab_bioinformatics.enrichment.gsea2005.GSEA2005">[docs]</a>
<span class="k">def</span> <span class="nf">GSEA2005</span><span class="p">(</span><span class="n">geneset_membership</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">correlations</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
<span class="w">  </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">  Implementation of algorithm described here:</span>
<span class="sd">  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/</span>

<span class="sd">  :param geneset_membership: (pd.Series) True if in set, False if not, index: all genes</span>
<span class="sd">  :param correlations: (pd.Series) Correlation of a given gene</span>
<span class="sd">  :return (Tuple[np.array, np.array]) x and y arrays ready to be plotted. ES = y.max()</span>
<span class="sd">  &#39;&#39;&#39;</span>
  <span class="n">r_j</span> <span class="o">=</span> <span class="n">correlations</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1"># r_j: correlation of gene_j in ranked order</span>
  <span class="n">S</span> <span class="o">=</span> <span class="n">geneset_membership</span><span class="p">[</span><span class="n">correlations</span><span class="o">.</span><span class="n">index</span><span class="p">]</span> <span class="c1"># S: geneset mask aligned with r_j</span>
  <span class="n">N</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>             <span class="c1"># N: number of genes</span>
  <span class="n">N_H</span> <span class="o">=</span> <span class="n">S</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>             <span class="c1"># N_H: number of hits</span>
  <span class="n">N_R</span> <span class="o">=</span> <span class="n">r_j</span><span class="p">[</span><span class="n">S</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>        <span class="c1"># N_R: sum of r_j for g_j \in S</span>
  <span class="n">P_hit</span> <span class="o">=</span>    <span class="n">S</span>  <span class="o">*</span> <span class="n">r_j</span><span class="o">/</span><span class="n">N_R</span>   <span class="c1"># P_hit: fraction of hits weighted by r_j</span>
  <span class="n">P_miss</span> <span class="o">=</span> <span class="p">(</span><span class="o">~</span><span class="n">S</span><span class="p">)</span> <span class="o">*</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="n">N</span><span class="o">-</span><span class="n">N_H</span><span class="p">)</span> <span class="c1"># P_hit: fraction of misses up to position i</span>
  <span class="c1"># 0 added to beginning for plotting, doesn&#39;t affect sum</span>
  <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">N</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
  <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([[</span><span class="mi">0</span><span class="p">],</span><span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">P_hit</span> <span class="o">-</span> <span class="n">P_miss</span><span class="p">)])</span>
  <span class="k">return</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span></div>

</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">maayanlab_bioinformatics.enrichment.gsea2005</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2020, Ma&#39;ayanlab.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    </div>
  </body>
</html>