
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>maayanlab_bioinformatics.dge.characteristic_direction &#8212; Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nature.css" />
    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">maayanlab_bioinformatics.dge.characteristic_direction</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for maayanlab_bioinformatics.dge.characteristic_direction</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">chi2</span>
<span class="kn">from</span> <span class="nn">scipy.stats.mstats</span> <span class="kn">import</span> <span class="n">zscore</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>

<span class="c1"># TODO: revamp _chdir</span>
<span class="k">def</span> <span class="nf">_chdir</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">sampleclass</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">gamma</span><span class="o">=</span><span class="mf">1.</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">calculate_sig</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">nnull</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">sig_only</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">norm_vector</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
  <span class="sd">&quot;&quot;&quot; repurposed from https://github.com/MaayanLab/geode/blob/master/geode/geode.py#L10-L115</span>

<span class="sd">  Calculate the characteristic direction for a gene expression dataset</span>
<span class="sd">  </span>
<span class="sd">  Input:</span>
<span class="sd">    data: numpy.array, is the data matrix of gene expression where rows correspond to genes and columns correspond to samples</span>
<span class="sd">    sampleclass: list or numpy.array, labels of the samples, it has to be consist of 0, 1 and 2, with 0 being columns to be excluded, 1 being control and 2 being perturbation</span>
<span class="sd">        example: sampleclass = [1,1,1,2,2,2]</span>
<span class="sd">    genes: list or numpy.array, row labels for genes </span>
<span class="sd">    gamma: float, regulaized term. A parameter that smooths the covariance matrix and reduces potential noise in the dataset</span>
<span class="sd">    sort: bool, whether to sort the output by the absolute value of chdir</span>
<span class="sd">    calculate_sig: bool, whether to calculate the significance of characteristic directions</span>
<span class="sd">    nnull: int, number of null characteristic directions to calculate for significance</span>
<span class="sd">    sig_only: bool, whether to return only significant genes; active only when calculate_sig is True</span>
<span class="sd">    norm_vector: bool, whether to return a characteristic direction vector normalized to unit vector</span>
<span class="sd">  Output:</span>
<span class="sd">    A list of tuples sorted by the absolute value in descending order characteristic directions of genes.</span>
<span class="sd">      If calculate_sig is set to True, each tuple contains a third element which is the ratio of characteristic directions to null ChDir</span>
<span class="sd">  &quot;&quot;&quot;</span>
  
  <span class="c1">## check input</span>
  <span class="n">data</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
  <span class="n">sampleclass</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">sampleclass</span><span class="p">)))</span>
  <span class="c1"># masks</span>
  <span class="n">m_non0</span> <span class="o">=</span> <span class="n">sampleclass</span> <span class="o">!=</span> <span class="mi">0</span>
  <span class="n">m1</span> <span class="o">=</span> <span class="n">sampleclass</span><span class="p">[</span><span class="n">m_non0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
  <span class="n">m2</span> <span class="o">=</span> <span class="n">sampleclass</span><span class="p">[</span><span class="n">m_non0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>

  <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">gamma</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gamma has to be a numeric number&quot;</span><span class="p">)</span>
  <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">sampleclass</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">])</span> <span class="ow">and</span> <span class="nb">set</span><span class="p">(</span><span class="n">sampleclass</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">]):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;sampleclass has to be a list whose elements are in only 0, 1 or 2&quot;</span><span class="p">)</span>
  <span class="c1"># if m1.sum()&lt;2 or m2.sum()&lt;2:</span>
  <span class="c1">#   raise ValueError(&quot;Too few samples to calculate characteristic directions&quot;)</span>
  <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span> <span class="o">!=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of genes does not match the demension of the expression matrix&quot;</span><span class="p">)</span>

  <span class="c1">## normalize data</span>
  <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span> <span class="n">m_non0</span><span class="p">]</span>
  <span class="n">data</span> <span class="o">=</span> <span class="n">zscore</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="c1"># standardize for each genes across samples</span>

  <span class="c1">## start to compute</span>
  <span class="n">n1</span> <span class="o">=</span> <span class="n">m1</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="c1"># number of controls</span>
  <span class="n">n2</span> <span class="o">=</span> <span class="n">m2</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="c1"># number of experiments</span>

  <span class="c1">## the difference between experiment mean vector and control mean vector.</span>
  <span class="n">meanvec</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="n">m2</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">-</span> <span class="n">data</span><span class="p">[:,</span><span class="n">m1</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> 

  <span class="c1">## initialize the pca object</span>
  <span class="n">pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
  <span class="n">pca</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

  <span class="c1">## compute the number of PCs to keep</span>
  <span class="n">cumsum</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span> <span class="c1"># explained variance of each PC</span>
  <span class="n">keepPC</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cumsum</span><span class="p">[</span><span class="n">cumsum</span> <span class="o">&gt;</span> <span class="mf">0.001</span><span class="p">])</span> <span class="c1"># number of PCs to keep</span>

  <span class="n">v</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">components_</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">keepPC</span><span class="p">]</span><span class="o">.</span><span class="n">T</span> <span class="c1"># rotated data </span>
  <span class="n">r</span> <span class="o">=</span> <span class="n">pca</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">T</span><span class="p">)[:,</span><span class="mi">0</span><span class="p">:</span><span class="n">keepPC</span><span class="p">]</span> <span class="c1"># transformed data</span>

  <span class="n">dd</span> <span class="o">=</span> <span class="p">(</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="n">m1</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">r</span><span class="p">[</span><span class="n">m1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="n">m2</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span><span class="n">r</span><span class="p">[</span><span class="n">m2</span><span class="p">])</span> <span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">n1</span><span class="o">+</span><span class="n">n2</span><span class="o">-</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># covariance</span>
  <span class="n">sigma</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">dd</span><span class="p">))</span> <span class="c1"># the scalar covariance</span>

  <span class="n">shrunkMats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">inv</span><span class="p">(</span><span class="n">gamma</span><span class="o">*</span><span class="n">dd</span> <span class="o">+</span> <span class="n">sigma</span><span class="o">*</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">gamma</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="n">keepPC</span><span class="p">))</span>

  <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">meanvec</span><span class="p">),</span> <span class="n">shrunkMats</span><span class="p">))</span>

  <span class="k">if</span> <span class="n">norm_vector</span><span class="p">:</span>
    <span class="n">b</span> <span class="o">/=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">b</span><span class="p">)</span> <span class="c1"># normalize b to unit vector</span>

  <span class="n">grouped</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">item</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">b</span><span class="p">],</span><span class="n">b</span><span class="p">,</span><span class="n">genes</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">sort</span><span class="p">:</span>
    <span class="n">grouped</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">grouped</span><span class="p">,</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


  <span class="k">if</span> <span class="ow">not</span> <span class="n">calculate_sig</span><span class="p">:</span> <span class="c1"># return sorted b and genes.</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">[(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">grouped</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">res</span>
  <span class="k">else</span><span class="p">:</span> <span class="c1"># generate a null distribution of chdirs</span>
    <span class="n">nu</span> <span class="o">=</span> <span class="n">n1</span> <span class="o">+</span> <span class="n">n2</span> <span class="o">-</span> <span class="mi">2</span>
    <span class="n">y1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multivariate_normal</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">keepPC</span><span class="p">),</span> <span class="n">dd</span><span class="p">,</span> <span class="n">nnull</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">nu</span> <span class="o">/</span> <span class="n">chi2</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">nu</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="n">nnull</span><span class="p">))</span>
    <span class="n">y2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multivariate_normal</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">keepPC</span><span class="p">),</span> <span class="n">dd</span><span class="p">,</span> <span class="n">nnull</span><span class="p">)</span><span class="o">.</span><span class="n">T</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">nu</span> <span class="o">/</span> <span class="n">chi2</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">nu</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="n">nnull</span><span class="p">))</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">y2</span> <span class="o">-</span> <span class="n">y1</span> <span class="c1">## y is the null of v</span>

    <span class="n">nullchdirs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">y</span><span class="o">.</span><span class="n">T</span><span class="p">:</span>
      <span class="n">bn</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">v</span><span class="p">,</span><span class="n">shrunkMats</span><span class="p">),</span> <span class="n">v</span><span class="o">.</span><span class="n">T</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">col</span><span class="p">,</span><span class="n">v</span><span class="o">.</span><span class="n">T</span><span class="p">))</span>
      <span class="n">bn</span> <span class="o">/=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">bn</span><span class="p">)</span>
      <span class="n">bn</span> <span class="o">=</span> <span class="n">bn</span> <span class="o">**</span> <span class="mi">2</span>
      <span class="n">bn</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
      <span class="n">bn</span> <span class="o">=</span> <span class="n">bn</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="c1">## sort in decending order</span>
      <span class="n">nullchdirs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bn</span><span class="p">)</span>

    <span class="n">nullchdirs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">nullchdirs</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">nullchdirs</span> <span class="o">=</span> <span class="n">nullchdirs</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">b_s</span> <span class="o">=</span> <span class="n">b</span> <span class="o">**</span> <span class="mi">2</span> 
    <span class="n">b_s</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
    <span class="n">b_s</span> <span class="o">=</span> <span class="n">b_s</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="c1"># sorted b in decending order</span>
    <span class="n">relerr</span> <span class="o">=</span> <span class="n">b_s</span> <span class="o">/</span> <span class="n">nullchdirs</span> <span class="c1">## relative error</span>
    <span class="c1"># ratio_to_null</span>
    <span class="n">ratios</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">relerr</span><span class="p">)</span><span class="o">/</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">relerr</span><span class="p">)</span><span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mf">1.</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">meanvec</span><span class="p">),</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">meanvec</span><span class="p">))</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">[(</span><span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">item</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">ratio</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span><span class="p">,</span> <span class="n">ratio</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">grouped</span><span class="p">,</span> <span class="n">ratios</span><span class="p">)]</span> 
    <span class="c1"># print(&#39;Number of significant genes: %s&#39;%(np.argmax(ratios)+1))</span>
    <span class="k">if</span> <span class="n">sig_only</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">res</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">ratios</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
      <span class="k">return</span> <span class="n">res</span>

<div class="viewcode-block" id="characteristic_direction"><a class="viewcode-back" href="../../../maayanlab_bioinformatics.dge.html#maayanlab_bioinformatics.dge.characteristic_direction.characteristic_direction">[docs]</a><span class="k">def</span> <span class="nf">characteristic_direction</span><span class="p">(</span><span class="n">controls_mat</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">cases_mat</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">gamma</span><span class="o">=</span><span class="mf">1.</span><span class="p">,</span> <span class="n">nnull</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">norm_vector</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">calculate_sig</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
  <span class="sd">&#39;&#39;&#39; Given two separate dataframes (controls, cases) with a shared index (genes), we compute the characteristic direction coefficients for all genes.</span>
<span class="sd">  e.g.</span>

<span class="sd">  control_mat:</span>
<span class="sd">        |control_replicate_1|control_replicate_2|...</span>
<span class="sd">  gene_1|        ..         |        ..         |...</span>
<span class="sd">  gene_2|        ..         |        ..         |...</span>

<span class="sd">  cases_mat:</span>
<span class="sd">        |case_replicate_1|case_replicate_2|...</span>
<span class="sd">  gene_1|     ..         |     ..         |...</span>
<span class="sd">  gene_2|     ..         |     ..         |...</span>
<span class="sd">  &#39;&#39;&#39;</span>
  <span class="k">assert</span> <span class="p">(</span><span class="n">controls_mat</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">cases_mat</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(),</span> <span class="s1">&#39;Index between controls and cases must be the same&#39;</span>
  <span class="n">n_genes</span> <span class="o">=</span> <span class="n">controls_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
  <span class="c1"># Compute characteristic direction</span>
  <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="n">_chdir</span><span class="p">(</span>
      <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">controls_mat</span><span class="p">,</span> <span class="n">cases_mat</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
      <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">controls_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">*</span><span class="n">cases_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
      <span class="p">),</span>
      <span class="c1"># genes</span>
      <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">controls_mat</span><span class="o">.</span><span class="n">index</span><span class="p">)),</span>
      <span class="n">gamma</span><span class="o">=</span><span class="n">gamma</span><span class="p">,</span>
      <span class="n">nnull</span><span class="o">=</span><span class="n">nnull</span><span class="p">,</span>
      <span class="n">norm_vector</span><span class="o">=</span><span class="n">norm_vector</span><span class="p">,</span>
      <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
      <span class="n">calculate_sig</span><span class="o">=</span><span class="n">calculate_sig</span><span class="p">,</span>
    <span class="p">),</span>
    <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CD-coefficient&#39;</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">,</span> <span class="s1">&#39;Significance&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">calculate_sig</span> <span class="k">else</span> <span class="p">[</span><span class="s1">&#39;CD-coefficient&#39;</span><span class="p">,</span> <span class="s1">&#39;Index&#39;</span><span class="p">],</span>
  <span class="p">)</span>
  <span class="n">results</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Index&#39;</span><span class="p">]</span>
  <span class="n">results</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">controls_mat</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span>
  <span class="k">if</span> <span class="n">sort</span><span class="p">:</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;CD-coefficient&#39;</span><span class="p">)</span>
  <span class="k">return</span> <span class="n">results</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;Index&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></div>

<div class="viewcode-block" id="up_down_from_characteristic_direction"><a class="viewcode-back" href="../../../maayanlab_bioinformatics.dge.html#maayanlab_bioinformatics.dge.characteristic_direction.up_down_from_characteristic_direction">[docs]</a><span class="k">def</span> <span class="nf">up_down_from_characteristic_direction</span><span class="p">(</span><span class="n">expr</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">600</span><span class="p">):</span>
  <span class="sd">&#39;&#39;&#39; Using the output of `characteristic_direction`, we can extract the top n genes</span>
<span class="sd">  with the highest absolute characteristic direction coefficients and split them into `up` and `down`.</span>
<span class="sd">  &#39;&#39;&#39;</span>
  <span class="n">highest_abs_expr</span> <span class="o">=</span> <span class="n">expr</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">expr</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;CD-coefficient&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)[:</span><span class="n">top_n</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
  <span class="k">return</span> <span class="nb">type</span><span class="p">(</span><span class="s1">&#39;UpDownGeneset&#39;</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">(),</span> <span class="nb">dict</span><span class="p">(</span>
    <span class="n">up</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">highest_abs_expr</span><span class="p">[</span><span class="n">highest_abs_expr</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
    <span class="n">down</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">highest_abs_expr</span><span class="p">[</span><span class="n">highest_abs_expr</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
  <span class="p">))</span></div>
</pre></div>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">Ma&#39;ayanlab Bioinformatics 0.0.1 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">maayanlab_bioinformatics.dge.characteristic_direction</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Ma&#39;ayanlab.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.0.2.
    </div>
  </body>
</html>